File: tarcreater.r
#!/usr/bin/env Rscript

if(!require("pacman")) install.packages("pacman")
pacman::p_load_gh("cmap/cmapR")

args <- list()

set_arguments <- function() {
  sysargs <- commandArgs(trailingOnly = TRUE)
  sysalen <- length(sysargs)
  valid   <- c('-p', '-e', '-o', '-m')
  index   <- 1
  while (index <= sysalen && sysargs[index] != '-m'){
    option <- sysargs[index]
    if (option %in% valid){
      args[[option]] <<- sysargs[index+1]
      index <- index + 2
    } else if (!(option %in% valid)){
      stop("# Error: Invalid Option.\n", call. = TRUE)
    }
  }
  index <- index + 1
  while (index <= sysalen){
    args[["-m"]] <<- c(args[["-m"]], sysargs[index])
    index <- index + 1
  }
  if (!("-p" %in% names(args)) || !("-o" %in% names(args))){
    stop("# Error: -p/-o required.\n", call. = TRUE)
  }
}

extract_expt_desn <- function(){
  return(read.csv(args[["-e"]], header = TRUE, sep = ','))
}

check_gct_against_expt <- function(myfile){
  out <- parse.gctx(myfile)
  if (out@version == "#1.2"){
    if (!("-e" %in% args)){
      stop("# Error: -e option required for .GCT (1.2).\n", call. = TRUE)
    }
  }
  gctcid <- out@cid
  expcsv <- extract_expt_desn()
  if (nrow(expcsv) < length(gctcid)){
    stop("# Error: Samles in .GCT not found in EXPT DESN file.\n", call. = TRUE)
  } else if (nrow(expcsv) > length(gctcid)){
    print(paste0("# Warning: Some Samples Missing in ", myfile, " file."))
  } else {
    print("# Success: Check OK.\n")
  }
}

stub <- function(myfile){
  
}

check_format <- function(myfile){
  ext <- strsplit(myfile, "[.]")[[1]][-1]
  opt <- list("gct"=check_gct_against_expt, "cct"=stub)
  opt[[ext]](myfile)
}

create_tar <- function(){
  option <- strsplit(args[["-o"]], "[/]")[[1]]
  mydirs <- head(option[-1], -1)
  anadir <- ""
  for (dir in mydirs){
    anadir <- file.path(anadir, dir)
  }
  tardir <- file.path(anadir, strsplit(tail(option, 1), "[.]")[[1]][1])
  system(paste0("mkdir -p ", tardir))
  system(paste0("mkdir -p ", tardir, "/data"))
  system(paste0("mkdir -p ", tardir, "/parsed-data"))
  system(paste0("cp ", args[["-e"]], " ", tardir, "/data/."))
  system(paste0("cp ", args[["-p"]], " ", tardir, "/parsed-data/."))
  if ("-m" %in% names(args)){
    for (extra in args[["-m"]]){
      system(paste0("cp ", extra, " ", tardir, "/parsed-data/."))
    }
  }
  system(paste0("tar -cvf ", args[["-o"]], " ", tardir))
  system(paste0("rm -rf ", tardir))
}

set_arguments()
check_format(args[["-p"]])
create_tar()

File: src/tarcreater.py
import execute_job
import sys
import os
import pdb
import cmapPy.pandasGEXpress.parse as parse_gct

os.environ["TF_CPP_MIN_LOG_LEVEL"]="3"
args = {}

def set_arguments():
    global set_arguments
    valid = ['p', 'e', 'o', 'm']
    arglen = len(sys.argv)
    index = 1
    while (index < arglen and sys.argv[index][1:] != 'm'):
        option = sys.argv[index]
        if option[0] == '-' and option[1:] in valid:
            args[option[1:]] = sys.argv[index+1]
            index += 2
        elif option[0] == '-' and option[1:] not in valid:
            print("# Error: Invalid Option.\n")
            usage()
            sys.exit(1)
        else:
            pass
    while (index < arglen):
        if 'm' not in args:
            args['m'] = []
        else:
            args['m'].append(sys.argv[index])
        index += 1
    if 'p' not in args or 'o' not in args:
        print("\n# ERROR: -p/-o required.")
        usage()
        sys.exit(1)


def extract_expt_desn():
    fp = open(args['e'], 'r')
    meta = fp.readline().strip().split(',')
    samples = fp.readlines()
    cdesc = {}
    for sample in samples:
        sample = sample.strip().split(',')
        cdesc[sample[0]] = sample[1:]
    fp.close()
    return cdesc, meta

def usage():
    print("\nUsage: python main.py \
    \n\t -p <ome-file> \
    \n\t -e <experiment-design-file> \
    \n\t -o <output tar file> \
    \n\t [-m (optional field) <extra_file_1> <extra_file_2> ...]")

def create_tar():
    tardir = args['o'].split('.')[0]
    commParamArr = []
    commParamArr.append([ 'mkdir', tardir ])
    commParamArr.append([ 'mkdir', tardir + "/data" ])
    commParamArr.append([ 'mkdir', tardir + "/parsed-data" ])
    commParamArr.append([ 'cp', args['e'], tardir + "/data/" ])
    commParamArr.append([ 'cp', args['p'], tardir + "/parsed-data/" ])
    if 'm' in args:
        for extra in args['m']:
            commParamArr.append(['cp'] + [extra] + [tardir + "/parsed-data/"])
    commParamArr.append([ 'tar', '-cvf', args['o'], tardir ])
    commParamArr.append([ 'rm', '-rf', tardir])
    for commParams in commParamArr:
        allok = execute_job.execute_job(commParams)
        if not allok:
            print("\n# ERROR: Aborting this program.")
            sys.exit(1)

def check_gct_against_expt(myfile):
    out = parse_gct.parse(myfile)
    print(out.version)
    if out.version == "GCT1.2":
        if 'e' not in args:
            usage()
            sys.exit(1)
    cid = out.col_metadata_df.index
    ecid, emeta = extract_expt_desn()
    if len(ecid) < len(cid):
        print("# Error: Samples in .GCT not found in EXPT DESN file.")
        sys.exit(1)
    elif len(ecid) > len(cid):
        print "# WARNING: Some Samples Missing in", myfile, "file."
    else:
        pass

def stub(myfile):
    pass

def check_format(myfile):
    ext = myfile.split('.')[-1]
    options = {
        'gct' : check_gct_against_expt,
        'cct' : stub
    }
    options[ext](myfile)

def main():
    pdb.set_trace()
    set_arguments()
    check_format(args['p'])
    create_tar()

if __name__ == "__main__":
    main()
